subprojects/packagefiles/libarchive/meson.build

```diff
@@ -299,16 +299,42 @@ else
   cdata.set('ICONV_CONST', cc.compiles(iconv_const_test) ? 'const' : '')
 endif
 
-regex_dep = disabler()
-pcre_dep = dependency('libpcre', disabler: true, required: get_option('regex'))
-if pcre_dep.found()
-  deps += pcre_dep
-elif get_option('regex').auto()
-  regex_dep = cc.find_library('regex', required: false)
-  if cc.has_function('regcomp', dependencies: regex_dep)
-    deps += regex_dep
+# If `regex` is support is enabled, check for
+# - glibc support first
+# - `libregex` second
+# - `libpcreposix` last
+rx_support = false
+rx_dep = get_option('regex').disabled() ? declare_dependency() : disabler()
+if not rx_dep.found()
+  if cc.has_header('regex.h') and cc.has_function('regcomp')
+    rx_dep = declare_dependency()
+    cdata.set('HAVE_REGEX_H', 1)
+    rx_support = 'glibc'
   endif
 endif
+if not rx_dep.found()
+  regex_dep = cc.find_library('regex', disabler: true, required: false)
+  if cc.has_header_symbol('regex.h', 'regcomp', dependency: regex_dep.partial_dependency(compile_args: true, includes: true))
+    rx_dep = regex_dep
+    cdata.set('HAVE_REGEX_H', 1)
+    rx_support = 'libregex'
+  endif
+endif
+if not rx_dep.found()
+  pcre_dep = dependency('libpcreposix', required: false)
+  if cc.has_header('pcreposix.h', dependencies: pcre_dep.partial_dependency(compile_args: true, includes: true))
+    rx_dep = pcre_dep
+    cdata.set('HAVE_PCREPOSIX_H', 1)
+    rx_support = 'libpcreposix'
+  endif
+endif
+if not rx_dep.found()
+  if get_option('regex').enabled()
+    error('Feature regex cannot be enabled')
+  endif
+  rx_dep = declare_dependency()
+endif
```

Comment:
But I think the error message is more confusing for the user.

Reply:
I can hear that logic.

Do you think that the `not_found_message: 'some message'` kwarg to dependency() helps here? It is printed as `message()` immediately after the dependency if it isn't found -- and then if it's required, the failure is re-raised. It would looks like this:

```
Run-time dependency foobar found: NO (tried pkgconfig and cmake)
Message: this is a not found message

meson.build:3:0: ERROR: Dependency "foobar" not found, tried pkgconfig and cmake
```

Or

```
Dependency foobar skipped: feature foo disabled
Message: this is a not found message
Build targets in project: 1

Found ninja-1.11.0.git at /usr/local/bin/ninja
```

Reply:
I think it's still less readable (not the last message, not an error), plus in this case the possibility of missing `pcreposix.h` still need to be accounted for. Although the macOS CI build is currently happy with the glibc provided support, but I have no idea if older / other macOS version can be assumed to also provide it.

Reply:
All right then.